// cozydb/lib/utils/later_stream.js
// Generated by CoffeeScript 1.8.0
(function() {
  var EventEmitter, LaterStream, drainStream,
    __bind = function(fn, me){ return function(){ return fn.apply(me, arguments); }; },
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  EventEmitter = require('events').EventEmitter;

  drainStream = function(stream, cb) {
    var body;
    body = '';
    stream.on('data', function(chunk) {
      if (cb) {
        return body += chunk;
      }
    });
    return stream.on('end', function() {
      return typeof cb === "function" ? cb(body) : void 0;
    });
  };

  module.exports = LaterStream = (function(_super) {
    __extends(LaterStream, _super);

    function LaterStream(callback) {
      this.callback = callback;
      this.onReadableReady = __bind(this.onReadableReady, this);
      this.pipe = __bind(this.pipe, this);
      this.abort = __bind(this.abort, this);
      this._onStreamingDone = __bind(this._onStreamingDone, this);
      LaterStream.__super__.constructor.apply(this, arguments);
      this.pipeDests = [];
      this.aborted = false;
      this.callbackCalled = false;
      this.trueStream = null;
    }

    LaterStream.prototype._onStreamingDone = function(err) {
      if (!this.callbackCalled) {
        this.callbackCalled = true;
        return this.callback(err);
      }
    };

    LaterStream.prototype.abort = function() {
      if (this.trueStream) {
        return this.trueStream.req.abort();
      } else {
        return this.aborted = true;
      }
    };

    LaterStream.prototype.pipe = function(dest) {
      if (this.trueStream) {
        if (typeof this.pipefilter === "function") {
          this.pipefilter(this.trueStream, dest);
        }
        return this.trueStream.pipe(dest);
      } else {
        return this.pipeDests.push(dest);
      }
    };

    LaterStream.prototype.onReadableReady = function(error, stream) {
      var dest, _i, _len, _ref, _results;
      if (error) {
        drainStream(stream);
        return this._onStreamingDone(error);
      } else if ((stream != null ? stream.statusCode : void 0) !== 200) {
        return drainStream(stream, (function(_this) {
          return function(body) {
            error = new Error("Error code " + (stream != null ? stream.statusCode : void 0) + " - " + body);
            error.status = (stream != null ? stream.statusCode : void 0) || 500;
            return _this._onStreamingDone(error);
          };
        })(this));
      } else if (this.aborted) {
        stream.req.abort();
        return drainStream(stream);
      } else {
        this.trueStream = stream;
        this.emit('ready', this.trueStream);
        this.trueStream.on('error', this._onStreamingDone);
        this.trueStream.on('end', this._onStreamingDone);
        _ref = this.pipeDests;
        _results = [];
        for (_i = 0, _len = _ref.length; _i < _len; _i++) {
          dest = _ref[_i];
          if (typeof this.pipefilter === "function") {
            this.pipefilter(this.trueStream, dest);
          }
          _results.push(this.trueStream.pipe(dest));
        }
        return _results;
      }
    };

    return LaterStream;

  })(EventEmitter);

}).call(this);
